fix: Search/Query may failed during updating delegator cache.

weiliu1031 opened the pull request:
issue: #37115

casue init query node client is too heavy, so we remove updateShardClient from leader mutex, which cause much more concurrent cornor cases.

This PR delay query node client's init operation until `getClient` is called, then use leader mutex to protect updating shard client progress to avoid concurrent issues.

weiliu1031 commented:
/run-cpu-e2e

xiaofan-luan commented on internal/proxy/meta_cache.go:
```diff
@@ -1016,6 +1007,15 @@ func (m *MetaCache) GetShards(ctx context.Context, withCache bool, database, col
 	// and create new client for new leaders
 	_ = m.shardMgr.UpdateShardLeaders(oldLeaders, ret)
```
please remove the UpdateShardLeaders function and init the shartClient in getClient

weiliu1031 replied:
refine the shard cache in next pr

weiliu1031 commented:
/hold

weiliu1031 committed:
fix: Search/Query may failed during updating delegator cache

casue init query node client is too heavy, so we remove
updateShardClient from leader mutex, which cause much more concurrent
cornor cases.

This PR delay query node client's init operation until `getClient` is
called, then use leader mutex to protect updating shard client progress
to avoid concurrent issues.

Signed-off-by: Wei Liu <wei.liu@zilliz.com>

weiliu1031 commented:
/unhold

weiliu1031 committed:
fix data race

Signed-off-by: Wei Liu <wei.liu@zilliz.com>

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 committed:
fix review comment

Signed-off-by: Wei Liu <wei.liu@zilliz.com>

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
rerun ut

weiliu1031 commented:
/run-cpu-e2e

xiaofan-luan commented on internal/proxy/shard_client.go:
```diff
@@ -31,26 +31,37 @@ var errClosed = errors.New("client is closed")
+		n.Lock()
+		if !n.initialized.Load() {
+			if err := n.initClients(); err != nil {
+				return nil, err
```
release n.lock

weiliu1031 committed:
fix: proxy retry to get shard leader on unloaded collection

pr#37116 let proxy retry to get shard leader if error happens, which
cause if search/query on a unloaded collection, which will keep retrying
until ctx done.

This PR add error type check to skip retry on ErrCollectionLoaded.

Signed-off-by: Wei Liu <wei.liu@zilliz.com>

weiliu1031 committed:
fix: dead lock if query node crash during shard client init

Signed-off-by: Wei Liu <wei.liu@zilliz.com>

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
rerun go-sdk

weiliu1031 commented:
rerun ut

weiliu1031 commented:
rerun ut

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
/run-cpu-e2e

weiliu1031 commented:
/run-cpu-e2e

czs007 commented:
/approve
/lgtm
